pkgs/binary-merkle-tree-halo2/src/binary_merkle_tree.rs

```diff
@@ -0,0 +1,264 @@
+#![warn(dead_code)]
+use anyhow::{Context, Ok, Result};
+use halo2_base::utils::BigPrimeField;
+use pse_poseidon::Poseidon;
+use rayon::{iter::ParallelIterator, slice::ParallelSlice};
+use serde::{Deserialize, Serialize};
+
+pub struct BinaryMerkleTree<'a, F, const T: usize, const RATE: usize, const ARITY: usize>
+where
+    F: BigPrimeField,
+{
+    pub root: F,
+    leaves: Vec<F>,
+    poseidon: &'a mut Poseidon<F, T, RATE>,
+    is_tree_ready: bool,
+    layers: Vec<Vec<F>>,
+    depth: usize,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct MerkleProofBytes {
+    pub siblings: Vec<u8>,
+    pub path_indices: Vec<u8>,
+    pub root: Vec<u8>,
+}
+
+impl MerkleProofBytes {
+    pub fn serialize(&self) -> Result<Vec<u8>> {
+        bincode::serialize(self).context("Failed to serialize MembershipProof")
+    }
+
+    pub fn deserialize(serialized: &[u8]) -> Result<Self> {
+        bincode::deserialize(serialized).context("Failed to deserialize MembershipProof")
+    }
+}
+
+pub struct MerkleProof<F>
+where
+    F: BigPrimeField,
+{
+    pub depth: usize,
+    pub leaf: F,
+    pub siblings: Vec<F>,
+    pub path_indices: Vec<usize>,
+    pub root: F,
+}
+
+impl<F> MerkleProof<F>
+where
+    F: BigPrimeField,
```

Comment:
```suggestion
impl<F: BigPrimeField> MerkleProof<F>
```

same as above, we have only one trait bound here, so my style preference is to define it inside the `<>` directly instead of `where`